### PR TITLE
add KafkaProducer post-processor to enable extensions from other libraries

### PR DESCRIPTION
**Pull request by sszp**
## Context

I'd like to enable async traces for messages sent out with `tw-tkms`.

To able to use OpenTelemetry's [KafkaTelemetry](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/67ba0a3fe618b48441a6255f7ab91f7f5bc9405c/instrumentation/kafka/kafka-clients/kafka-clients-2.6/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/v2_6/KafkaTelemetry.java#L53) class from the `tw-observability-base`, I'm adding a new `KafkaProducer` post-processor interface that can be used from `tw-observability-base` to call the `KafkaTelemetry:wrap()` to create the necessary async traces for all message sent out with `tw-tkms`.

## Checklist
- [x] Change meets or does not compromise the [Baseline Security Requirements](https://transferwise.atlassian.net/wiki/spaces/EKB/pages/434929973/Baseline+Security+Requirements)

### CODE REVIEW

File: tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsKafkaProducerProvider.java [hunk 1]
```diff
@@ -39,8 +42,12 @@ public class TkmsKafkaProducerProvider implements ITkmsKafkaProducerProvider, Gr
 
   private Map<Pair<TkmsShardPartition, UseCase>, ProducerEntry> producers = new ConcurrentHashMap<>();
 
+  @Autowired
+  private List<ITkmsKafkaProducerPostProcessor> postProcessors = new ArrayList<>();
```
Comment (normanma-tw) [inline tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsKafkaProducerProvider.java:46]: Do you think there's likely to be a situation where there will be multiple `ITkmsKafkaProducerPostProcessor`s in the Spring context and we're interested in defining their ordering? If the extensions are being provided by other libraries I'm not sure how we'd do that in the auto configuration.

**Comment by tw-peeterkarolin** [comment]
I think it would make sense to play this through with a snapshot version in combination with tw-observability-base to understand how the whole flow will work.
Just using `KafkaTelemetry:wrap()` won't be enough, because that that will work with normal use cases where the OpenTelemetry context is configured and available for the KafkaTelemetry wrapper to use. In tw-tkms polling the messages from the DB and publishing them to Kafka will happen in a separate poller thread, so it will need to initialise the OpenTelemetry context as well, before KafkaTelemetry can work.

**Comment by sszp** [comment]
> I think it would make sense to play this through with a snapshot version in combination with tw-observability-base to understand how the whole flow will work. Just using `KafkaTelemetry:wrap()` won't be enough, because that that will work with normal use cases where the OpenTelemetry context is configured and available for the KafkaTelemetry wrapper to use. In tw-tkms polling the messages from the DB and publishing them to Kafka will happen in a separate poller thread, so it will need to initialise the OpenTelemetry context as well, before KafkaTelemetry can work.

I plan to look into that. At first I plan to see if I can use `ITkmsMessageDecorator` to add headers to the message and get those headers from the message using `ITkmsMessageInterceptors` just before it's sent out using the Kafka Client to initialise the OpenTelemetry context before sending.

**Review by normanma-tw — DISMISSED** [review]
LGTM once the snapshot version is cleaned up + optionally update the date of the changelog

File: tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsAutoConfiguration.java [hunk 2]
```diff
@@ -51,4 +49,9 @@ public List<ITkmsMessageDecorator> messageDecorators() {
     return Collections.emptyList();
   }
 
+  @Bean
```
Comment (tw-peeterkarolin) [inline tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsAutoConfiguration.java:52]: Probably would be better to declare the autowire usage as
`@Autowired(required = false)`
Then you don't need to explicitly provide an empty list bean.

**Comment by sszp** [comment]
> LGTM once the snapshot version is cleaned up + optionally update the date of the changelog
That's done now. @normanma-tw

File: tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsAutoConfiguration.java (hunk 2 above)
Comment (sszp) [reply tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/config/TkmsAutoConfiguration.java]: Fixed that.

**Review by normanma-tw — APPROVED** [review]

**Review by tw-peeterkarolin — APPROVED** [review]

**Pull request merged**